src/test/java/rx/internal/operators/OperatorValveTest.java

```diff
@@ -0,0 +1,163 @@
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+        // next batch
+        order.verify(inputProducer).request(2);// request limited called from onNext
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+        // next batch
+        order.verify(inputProducer).request(1);
+        control.onNext(false);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+
+        // no request while closed 
+        control.onNext(true);
+        control.onNext(true);
+
+        // small request while closed 
+        control.onNext(false);
+        outputProducer.request(1);
+        control.onNext(true);
+        order.verify(inputProducer).request(1);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+
+        // larger request while closed 
+        control.onNext(false);
+        outputProducer.request(2);
+        control.onNext(true);
+        order.verify(inputProducer).request(2);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+
+        // too large of a request while closed 
+        control.onNext(false);
+        outputProducer.request(3);
+        control.onNext(true);
+        order.verify(inputProducer).request(2);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+        order.verify(inputProducer).request(1);
+        input.onNext(++i);
+        order.verify(output).onNext(i);
+
+        input.onCompleted();
+        order.verify(output).onCompleted();
+
+        // all done
+        order.verifyNoMoreInteractions();
+    }
+
+    @Test
+    public void testDataError() {
+        TestSubscriber<Void> tSub = TestSubscriber.create();
+        Exception e = new Exception();
+        Observable.<Void> error(e).pressureValve(Observable.<Boolean> never(), 10).subscribe(tSub);
+        tSub.assertError(e);
+    }
+
+    @Test
+    public void testControlError() {
+        TestSubscriber<Void> tSub = TestSubscriber.create();
+        Exception e = new Exception();
+        Observable.<Void> never().pressureValve(Observable.<Boolean> error(e), 10).subscribe(tSub);
+        tSub.assertError(e);
+    }
+
+    @Test
+    public void testControlCompleteOpen() {
+        TestSubscriber<Integer> tSub = TestSubscriber.create();
+        Observable.just(1, 2, 3).pressureValve(Observable.<Boolean> empty(), 10).subscribe(tSub);
+        tSub.assertValues(1, 2, 3);
+    }
+
+    @Test
+    public void testControlCompleteClosed() {
+        TestSubscriber<Void> tSub = TestSubscriber.create();
+        Observable.<Void> never().pressureValve(Observable.just(false), 10).subscribe(tSub);
+        tSub.assertError(IllegalStateException.class);
+    }
+}
```

Comment:
I'm having a hard time coming up with consistent unit tests for this. It is designed as a very sloppy back pressure mechanism; reactive and not prescriptive.  If I used the TestScheduler then there isn't much difference between the units tests you want and the contrived tests that I have now. For example
```java
    @Test
    public void testObserveOn() {
        final AtomicLong counter = new AtomicLong();
        Observable<Integer> range = Observable.range(0, Integer.MAX_VALUE);
        Observable<Boolean> control = Observable.interval(1, TimeUnit.SECONDS).map(new Func1<Long, Boolean>() {
            @Override
            public Boolean call(Long i) {
                System.out.println();
                counter.set(0);
                return i % 2 == 1;
            }
        });
        long granularity = 10;
        TestSubscriber<Integer> tSub = new TestSubscriber<Integer>();
        range.pressureValve(control, granularity).observeOn(Schedulers.computation()).toBlocking().forEach(new Action1<Integer>() {
            @Override
            public void call(Integer t) {
                System.out.print(counter.incrementAndGet()+ "         \r");
            }
        });
    }
```
Produces the output like this where the number of `onNext`s received after the valve is closed can be anywhere from 8 to 127.
```
550343         
56             
598760         
23             
593937         
110            
597200         
47             
587776         
127            
586648         
103            
585984         
127            
579781         
58             
580374         
105            
581879         
8              
587867         
36             
585960         
23             
572516         
27             
581530         
101            
579535         
48             
583840         
95             
579200         
127            
583312         
111            
580183         
40             
578295         
8              
597470         
```